sentry-core/src/performance.rs

```diff
@@ -0,0 +1,452 @@
+use std::sync::Arc;
+use std::sync::Mutex;
+
+use crate::protocol;
+use crate::{Client, Hub};
+
+const MAX_SPANS: usize = 1_000;
+
+// global API:
+
+/// Start a new Performance Monitoring Transaction.
+///
+/// The transaction needs to be explicitly finished via [`Transaction::finish`],
+/// otherwise it will be discarded.
+/// The transaction itself also represents the root span in the span hierarchy.
+/// Child spans can be started with the [`Transaction::start_child`] method.
+pub fn start_transaction(ctx: TransactionContext) -> Transaction {
+    let client = Hub::with_active(|hub| hub.client());
+    Transaction::new(client, ctx)
+}
+
+// Hub API:
+
+impl Hub {
+    /// Start a new Performance Monitoring Transaction.
+    ///
+    /// See the global [`start_transaction`] for more documentation.
+    pub fn start_transaction(&self, ctx: TransactionContext) -> Transaction {
+        Transaction::new(self.client(), ctx)
+    }
+}
+
+// "Context" Types:
+
+/// The Transaction Context used to start a new Performance Monitoring Transaction.
+///
+/// The Transaction Context defines the metadata for a Performance Monitoring
+/// Transaction, and also the connection point for distributed tracing.
+#[derive(Debug)]
+pub struct TransactionContext {
+    name: String,
+    op: String,
+    trace_id: protocol::TraceId,
+    parent_span_id: Option<protocol::SpanId>,
+    sampled: Option<bool>,
+}
```

Comment:
slightly nitpicky question; is there any reason why this couldn't be abstracted away into `Transaction` itself? what's the benefit of explicitly exposing `TransactionContext` as opposed to just having just `Transaction` by itself (presumably with all of `TransactionContext`'s methods on `Transaction` instead)? you could also move `start_transaction` so it becomes `Transaction::start()`, and then creating a transaction is a matter of just:

```
Transaction::new("honk", "beep");
# Transaction::set_sampled(...);
Transaction::start();
```

or 
```
Transaction::continue_from_headers(...);
Transaction::start();
```

Reply:
it also avoids interesting discrepancies between the two structs, such as `TransactionContext`'s `op` being `String` but `Transaction`'s `op` being `Option<String>`.

Reply:
Interesting. I don’t really want to go down the route of bikeshedding this.